### PR TITLE
Add installer task, refactor runner task 

### LINKED ISSUE #11: Find/create icon?
I feel like it would be good to include an icon with the task. The [VS Code Extension for ShellCheck](https://marketplace.visualstudio.com/items?itemName=timonwong.shellcheck) didn't add one, but for Azure Pipelines tasks it bothers me when there's no icon/default icon. 

The favicon used on the ShellCheck site looks solid to me, although would probably need to double check on licensing/permissions/attribution/etc. https://www.shellcheck.net/favicon.ico

image

@beverts312 and/or @traviskosarek - thoughts or preferences?

### LINKED ISSUE #13: Create ShellCheck Installer Task
Currently the run task will auto-install ShellCheck if the bin cannot be found. 

We should extract that into a separate tool installer task, which would give users more control over install behavior (including version selection)

Originally, I viewed this as more like a "nice to have" in my opinion vs. a "must-have for 1.x release", however, with the existing auto-install behavior the hosted agents are essentially caching shellcheck once installed which removes the users ability to specify version.

CC @beverts312

### PR DESCRIPTION
**Pull request by calebcartwright**
Close #11, close #13 

The dif in this PR looks bigger than it actually is. This is mostly just a refactor of existing code, which results in splitting out the capabilities of _installing_ and _running_ ShellCheck into two separate, standalone tasks along with some other minor refactoring (and it also adds icons for the tasks and extension). 

The functionality/implementation is basically the same, just refactored into separate files/functions. I've added some inline comments for reference.

Here's what the tasks look like in the UI (using the visual-designer):
image

The [sample pipeline](https://dev.azure.com/swellaby/OpenSource/_apps/hub/ms.vss-ciworkflow.build-ci-hub?_a=edit-build-definition&id=67) can also be referenced

### CODE REVIEW

File: test/unit/tasks/install/installer.ts [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+            osTypeStub.callsFake(() => 'linux');
+            osArchStub = Sinon.stub(os, 'arch');
+            pathJoinStub = Sinon.stub(path, 'join');
+            pathJoinStub.withArgs(tempDirectoryPath, binaryDirectoryName).callsFake(() => extractRoot);
+            pathJoinStub.withArgs(extractRoot, binaryDirectoryName).callsFake(() => binTargetDirectory);
+            toolLibDownloadStub.callsFake(() => Promise.resolve(downloadDirectory));
+            taskLibGetVariableStub = Sinon.stub(taskLib, 'getVariable');
+            taskLibGetVariableStub.withArgs('Agent.TempDirectory').callsFake(() => tempDirectoryPath);
+            taskLibMkdirPStub = Sinon.stub(taskLib, 'mkdirP');
+            // toolLibExtractTarStub = Sinon.stub(toolLib, 'extractTar');
```
Comment (beverts312) [inline test/unit/tasks/install/installer.ts]:
```suggestion
```

File: test/unit/tasks/install/installer.ts [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            // toolLibExtractTarStub.withArgs(downloadDirectory).callsFake(() => Promise.resolve(tempRoot));
```
Comment (beverts312) [inline test/unit/tasks/install/installer.ts]:
```suggestion
```

File: src/tasks/install/installer.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        tarballName = `shellcheck-${version}.linux.armv6hf.tar.xz`;
+    } else {
+        throw new Error(`Unsupported architecture ${architecture}`);
+    }
+
+    const downloadUrl = `${shellCheckBinaryUrlBase}/${tarballName}`;
+    const tarballLocation = await toolLib.downloadTool(downloadUrl);
+    // The ShellCheck binary tarballs are not gzip compressed. The toolLib.extractTar
+    // function always adds the 'z' tar option which would always fail for ShellCheck.
+    //const extractRoot = await toolLib.extractTar(tarballLocation);
```
Comment (beverts312) [inline src/tasks/install/installer.ts]:
```suggestion
```

File: test/unit/tasks/install/installer.ts [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    suite('Linux installer', () => {
+        const version = '0.6.0';
+        const binaryDirectoryName = `shellcheck-${version}`;
+        const downloadDirectory = '/foo/bar';
+        const tempDirectoryPath = '/vsts/work/_temp';
+        const extractRoot = `${tempDirectoryPath}/${binaryDirectoryName}`;
+        const binTargetDirectory = `${extractRoot}/${binaryDirectoryName}`;
+        let osArchStub: Sinon.SinonStub;
+        let pathJoinStub: Sinon.SinonStub;
+        // let toolLibExtractTarStub: Sinon.SinonStub;
```
Comment (beverts312) [inline test/unit/tasks/install/installer.ts]:
```suggestion
```

**Comment by calebcartwright** [comment]
> found some commented out code 

👍 Yeah good spot. That was the original implementation that I meant to take out. My hope is to send a PR to the tool installer lib to make gzip support optional via a param so that we can leverage the built in function. 
If/when that happens we can update this accordingly.

**Pull request merged**